src/types/utils.ts

```diff
@@ -296,15 +296,22 @@ type OptionalKeysOfHelper<
 /**
  * Filter out nevers from a tuple.
  */
-// TODO: Make tail-recursive.
 export type FilterOutNever<T extends ReadonlyArray<unknown>> =
-  T extends readonly []
-    ? []
-    : T extends [infer Head, ...infer Rest]
-    ? IsNever<Head> extends true
-      ? FilterOutNever<Rest>
-      : [Head, ...FilterOutNever<Rest>]
-    : T;
+  FilterOutNeverHelper<T, []>;
+
+/**
+ * Tail-recursive helper type for FilterOutNever.
+ */
+type FilterOutNeverHelper<
```

Comment:
I assume you want to get rid of `never` from the tuple. 
Test:
```typescript
type IsNever<T> = [T] extends [never] ? true : false

type FilterOutNeverHelper<
    T extends ReadonlyArray<unknown>,
    Acc extends ReadonlyArray<unknown>
    > = T extends readonly []
    ? Acc
    : T extends [infer Head, ...infer Rest]
    ? IsNever<Head> extends true
    ? FilterOutNeverHelper<Rest, Acc>
    : FilterOutNeverHelper<Rest, [...Acc, Head]>
    : T;

// readonly [1, 2, never, 4]
type Result = FilterOutNeverHelper<readonly [1, 2, never, 4], readonly []>
```
`Result` still contains `never` because `T extends [infer Head, ...infer Rest]` never evalueates to `true` since `[infer Head, ...infer Rest]` is mutable list. You need to use `readonly` flag also for this conditional type.

Consider this exmaple:
```typescript
type IsNever<T> = [T] extends [never] ? true : false

type FilterOutNeverHelper<
    T extends ReadonlyArray<unknown>,
    Acc extends ReadonlyArray<unknown>
    > = T extends readonly []
    ? Acc
    : T extends readonly [infer Head, ...infer Rest] // <---- change
    ? IsNever<Head> extends true
    ? FilterOutNeverHelper<Rest, Acc>
    : FilterOutNeverHelper<Rest, [...Acc, Head]>
    : T;

// readonly [1, 2, 4]
type Result = FilterOutNeverHelper<readonly [1, 2, never, 4], readonly []>
```

Reply:
Thanks. I usually don't miss that 😛

Reply:
Actually, when I make this change other parts of the code base now have type errors.

Reply:
The helper type will never be used from anywhere other than the main type, so it should be ok

Reply:
I don't have any errors. Where do you have them ?

Reply:
tests/deepmerge.test.ts(293,23)

Reply:
I'm not aware how you handle such cases and what the algorithm of handling `null`, hence it hard for me to help here

Reply:
Basically, if the property exists, no matter what is value is (be that `null` or `undefined`) it will replace the other value. `null` or `undefined` are not given special treatment.

Reply:
I have removed `readonly` from here `FilterOutNeverHelper<Rest, readonly [...Acc, Head]>` and seems it works

Reply:
That works for me too